File: apogee-backend/apogee/services/llm_service.py
import ollama


class LLMError(Exception):
    """Raised when the LLM backend is unreachable or returns an error."""
    pass


def generate_stream(prompt: str, model: str = "qwen3:8b"):
    """Yield tokens from Ollama. Raises LLMError on failure."""
    try:
        response = ollama.chat(
            model=model,
            messages=[
                {
                    "role": "user",
                    "content": prompt,
                }
            ],
            stream=True,
        )

        for chunk in response:
            yield chunk["message"]["content"]

    except ollama.ResponseError as exc:
        raise LLMError(
            f"Ollama returned an error for model '{model}': {exc}"
        ) from exc
    except Exception as exc:
        raise LLMError(
            f"Could not connect to Ollama. Is it running? Error: {exc}"
        ) from exc

File: apogee-backend/src/services/llmService.js
import ollama from "ollama";

export class LLMError extends Error {}

// The `ollama` package doesn't export its ResponseError class (only the
// client and its default instance), so API errors are distinguished by
// name/status_code duck-typing instead of `instanceof`.
function isOllamaResponseError(err) {
  return err?.name === "ResponseError" && "status_code" in err;
}

/** Async-generator yielding tokens from Ollama. Throws LLMError on failure. */
export async function* generateStream(prompt, model = "gemma3:4b") {
  let stream;
  try {
    stream = await ollama.chat({
      model,
      messages: [{ role: "user", content: prompt }],
      stream: true,
      think: false,
    });
  } catch (err) {
    throw wrapError(err, model);
  }

  try {
    for await (const chunk of stream) {
      yield chunk.message.content;
    }
  } catch (err) {
    throw wrapError(err, model);
  }
}

function wrapError(err, model) {
  if (isOllamaResponseError(err)) {
    return new LLMError(
      `Ollama returned an error for model '${model}': ${err.message}`,
    );
  }
  return new LLMError(
    `Could not connect to Ollama. Is it running? Error: ${err?.message ?? err}`,
  );
}

export async function generateText(prompt, model = "gemma3:4b") {
  let text = "";
  for await (const token of generateStream(prompt, model)) {
    text += token;
  }
  return text;
}
